sw/device/lib/dif/dif_spi_device.h

```diff
@@ -0,0 +1,85 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef SW_DEVICE_LIB_DIF_DIF_SPI_DEVICE_H_
+#define SW_DEVICE_LIB_DIF_DIF_SPI_DEVICE_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/dif/register.h"
+
+/**
+ * Configuration for initializing a SPI device.
+ */
+typedef struct dif_spi_device_config {
+  reg32_t base_addr;
+} dif_spi_device_config_t;
+
+/**
+ * State for a particular SPI device.
+ *
+ * Its member variables should be considered private, and are only provided so
+ * that callers can allocate it.
+ */
+typedef struct dif_spi_device { reg32_t base_addr; } dif_spi_device_t;
```

Comment:
Allocation.  If you don't have the complete type you can't allocate one (statically, preferably) somewhere else...

Reply:
Sure, I guess it depends on how much we want to hide this from clients. Two alternatives that may be worth considering (for static allocation):
 - A statically allocated instance (for single-instance cases) or pool (for multi-instance cases) in the DIF itself.
 - (Fragile and ugly, but ...) `#define SPI_DEVICE_SIZE XX` coupled with `static_assert(SPI_DEVICE_SIZE == sizeof(dif_spi_device_t), "...")` in `dif_spi_device.c`. Clients can allocate an array of bytes of size `SPI_DEVICE_SIZE` and cast it to `* dif_spi_device_t`.

Reply:
I do agree that making structs opaque to their clients is usually a good idea.

However, in the case of DIFs, we explicitly want DV to be extending the interface for their testing, so we should probably leave these structs "public", so that DV's DIFs can also rely on the same information. I don't forsee DV's DIFs needing to extend these structs, but I could be wrong.

I feel the right way to go about this is to catch problematic code in review, where "problematic code" is production non-DIF code accessing details of the DIF struct.